main/src/ca/uwaterloo/flix/api/lsp/provider/RenameProvider.scala

```diff
@@ -27,145 +30,517 @@ object RenameProvider {
+
+      // Extract trait symbol SourceLocation
+      val traitEndColWithoutSigName = sepCol.toShort
+      val traitSymEnd = SourcePosition(loc.sp2.source, loc.sp2.line, traitEndColWithoutSigName)
+      val traitSymLoc = SourceLocation(loc.isReal, loc.sp1, traitSymEnd)
+
+      // Extract sig symbol SourceLocation
+      val sigStartColWithoutTraitName = (sepCol + 1).toShort
+      val sigSymStart = SourcePosition(loc.sp1.source, loc.sp1.line, sigStartColWithoutTraitName)
+      val sigSymLoc = SourceLocation(loc.isReal, sigSymStart, loc.sp2)
+
+      (Some(traitSymLoc), sigSymLoc)
+    } else {
+      (None, loc)
+    }
   }
 
-  private def renameEnum(sym: Symbol.EnumSym, newName: String)(implicit index: Index): JObject = {
-    val defn = sym.loc
-    val uses = index.usesOf(sym)
-    rename(newName, uses + defn)
+  private def getStructFieldSymOccurs(sym: Symbol.StructFieldSym)(implicit root: Root): Set[SourceLocation] = {
+    var occurs: Set[SourceLocation] = Set.empty
+    def consider(s: Symbol.StructFieldSym, loc: SourceLocation): Unit = {
+      if (s == sym) { occurs += loc }
+    }
+    object StructFieldSymConsumer extends Consumer {
+      override def consumeStructField(field: TypedAst.StructField): Unit = consider(field.sym, field.sym.loc)
+      override def consumeStructFieldSymUse(symUse: SymUse.StructFieldSymUse): Unit = consider(symUse.sym, symUse.loc)
+    }
+
+    Visitor.visitRoot(root, StructFieldSymConsumer, AllAcceptor)
+
+    occurs
   }
 
-  private def renameTypeAlias(sym: Symbol.TypeAliasSym, newName: String)(implicit index: Index): JObject = {
-    val defn = sym.loc
-    val uses = index.usesOf(sym)
-    rename(newName, uses + defn)
+  private def getStructSymOccurs(sym: Symbol.StructSym)(implicit root: Root): Set[SourceLocation] = {
+    var occurs: Set[SourceLocation] = Set.empty
+    def consider(s: Symbol.StructSym, loc: SourceLocation): Unit = {
+      if (s == sym) { occurs += loc }
+    }
+    object StructSymConsumer extends Consumer {
+      override def consumeStruct(struct: TypedAst.Struct): Unit = consider(struct.sym, struct.sym.loc)
+      override def consumeType(tpe: Type): Unit = tpe match {
+        case Type.Cst(TypeConstructor.Struct(sym, _), loc) => consider(sym, loc)
+        case _ => ()
+      }
+    }
+
+    Visitor.visitRoot(root, StructSymConsumer, AllAcceptor)
+
+    occurs
   }
 
-  private def renameLabel(label: Name.Label, newName: String)(implicit index: Index): JObject = {
-    val defs = index.defsOf(label)
-    val uses = index.usesOf(label)
-    rename(newName, defs ++ uses)
+  private def getEnumSymOccurs(sym: Symbol.EnumSym)(implicit root: Root): Set[SourceLocation] = {
+    var occurs: Set[SourceLocation] = Set.empty
+    def consider(s: Symbol.EnumSym, loc: SourceLocation): Unit = {
+      if (s == sym) { occurs += loc }
+    }
+    object EnumSymConsumer extends Consumer {
+      override def consumeEnum(enm: TypedAst.Enum): Unit = consider(enm.sym, enm.sym.loc)
```

Comment:
We already have the enum symbol right? Here TypedAst.Enum is literally the same symbol -- its the declaration site.

Reply:
We just need the uses which we get from consumeCaseSymUse

Reply:
(This comment applies to many of the other functions too)

Reply:
Yes. But then we'd need a different `getOccurs` function for when the cursor is on the definition site vs. use site of a symbol for each symbol type.  I figured it it was simpler to have *one* function that gives the occurrences for each relevant symbol type and just repeat a little bit of search. But maybe it's better to to split into `getXOccurFromDeclSite` and `getXOccurFromUseSite` (or something like that) to avoid some of that search repetition?

Reply:
What do you think?

Reply:
I don't understand. We use on search to find the symbol and the one search to find all the SymUses. Why does the second search need to include the declaration?

Reply:
Oh, right, because when we've found the symbol, where the cursor is on a use of it or the declaration, we know the `SourceLocation` of the declaration, because it's baked into the symbol. So yeah, you're right. So in the second search we only ever need to find sym uses. I though that if we found the symbol through the cursor being on a SymUse, then we'd have to search for the declaration site, but ofc. we don't. So yeah, it can be simplified. Good cool :)

Reply:
Ping me when you have a refreshed PR

Reply:
Yep. Will do. I'll also be publishing a new PR soon with the FindReferencesProvider changes as well. Probably both will be done by tonight :)

Reply:
I see we are both burning the midnight oil

Reply:
Always XD